R/step_adjust_latency.R

```diff
@@ -0,0 +1,462 @@
+#' @export
+#' @importFrom glue glue
+#' @importFrom dplyr rowwise
+prep.step_adjust_latency <- function(x, training, info = NULL, ...) {
+  sign_shift <- get_sign(x)
+  latency <- x$latency
+  forecast_date <- x$forecast_date %||% set_forecast_date(training, info, x$epi_keys_checked, latency)
+  # construct the latency table
+  latency_table <- names(training)[!names(training) %in% key_colnames(training)] %>%
+    tibble(col_name = .)
+  if (length(recipes_eval_select(x$terms, training, info)) > 0) {
+    latency_table <- latency_table %>% filter(col_name %in%
+      recipes_eval_select(x$terms, training, info))
+  }
+
+  if (is.null(latency)) {
+    latency_table <- latency_table %>%
+      rowwise() %>%
+      mutate(latency = get_latency(training, forecast_date, col_name, sign_shift, x$epi_keys_checked))
+  } else if (length(latency) > 1) {
+    # if latency has a length, it must also have named elements. We assign based on comparing the name in the list
+    # with the column names, and drop any which don't have a latency assigned
+    latency_table <- latency_table %>%
+      filter(col_name %in% names(latency)) %>%
+      rowwise() %>%
+      mutate(latency = unname(latency[names(latency) == col_name])) %>%
+      ungroup()
+  } else {
+    latency_table <- latency_table %>% mutate(latency = latency)
+  }
+  attributes(training)$metadata$latency_table <- latency_table
+  # get the columns used, even if it's all of them
+  terms_used <- x$terms
+  # TODO replace with is_empty as in bake.recipe
+  if (is_empty(terms_used)) {
+    terms_used <- info %>%
+      filter(role == "raw") %>%
+      pull(variable)
+  }
+
+
+
+  step_adjust_latency_new(
+    terms = x$terms,
+    role = x$role,
+    trained = TRUE,
+    forecast_date = forecast_date,
+    latency = unique(latency_table$latency),
+    latency_table = latency_table,
+    metadata = attributes(training)$metadata,
+    keys = x$keys,
+    method = x$method,
+    epi_keys_checked = x$epi_keys_checked,
+    columns = recipes_eval_select(latency_table$col_name, training, info),
+    skip = x$skip,
+    id = x$id
+  )
+}
+
+#' @importFrom dplyr %>% pull group_by_at
+#' @importFrom tidyr fill
+#' @export
+bake.step_adjust_latency <- function(object, new_data, ...) {
+  if (!isa(new_data, "epi_df")) {
+    # TODO if new_data actually has keys other than geo_value and time_value, this is going to cause problems
+    new_data <- as_epi_df(new_data)
+    attributes(new_data)$metadata <- object$metadata
+    attributes(new_data)$metadata$as_of <- object$forecast_date
+  }
+  if (object$method == "extend_ahead" || object$method == "extend_lags") {
+    attributes(new_data)$metadata$latency_method <- object$method
+    attributes(new_data)$metadata$shift_sign <- get_sign(object)
+    attributes(new_data)$metadata$latency_table <- object$latency_table
+    attributes(new_data)$metadata$forecast_date <- object$forecast_date
+    keys <- object$keys
```

Comment:
What is the reasoning for adding all of this as attributes? We should talk about it I think.

Reply:
It's really the only way to pass the values along. I suppose only one of `latency_method` and `shift_sign` is actually necessary, but the latency table is definitely needed, along with forecast_date.

I guess it might be more appropriate to store this in the recipe, but I'm not really sure about that (or how to go about it).